### ISSUE TITLE
Strange/incorrect sphere shapes

### ISSUE THREAD

**Issue by scrapmech** [maruohon/minihud#215]
Minecraft version: 1.18.1
MiniHUD version: 0.21.3

When using Sphere block-based generator I noticed strange behavior: generated spheres are often not symmetrical, especially when using block snap corner or none options. Adjustable spawn sphere seems to produce better results.

Attached is an example for this behavior. You can see that on the right side there are several extra blocks.

![2022-02-15_20 11 05](https://user-images.githubusercontent.com/53707587/154119813-7a2c41ed-a342-4050-908f-08112b03fd5f.png)
![2022-02-15_20 11 13](https://user-images.githubusercontent.com/53707587/154119848-4debdf55-06fc-4065-b5ab-2fb8fd5c3cdf.png)

**Comment by maruohon** [comment]
I believe the "normal" sphere uses the distance of the block center to the center point as the deciding criteria for which blocks to consider as the edge. So if the center of the circle is not centered on a block, then I guess it's entirely possible that it will not look symmetrical because the "unsymmetrical distances" on the opposing sides of the sphere result in different blocks being selected as the edge.

**Comment by scrapmech** [comment]
@maruohon  The center of the sphere is exactly in the corner of a block though? The block centers of blocks that are symmetric are exactly the same distance from it. For example sphere with center at (0, 63, 0) and radius of 6. Two gold blocks whose centers are at (3.5, 66.5, 3.5) and (-3.5, 66.5, 3.5). The distance is sqrt((3.5^2)*3)=6.062177826. Left one is included in the sphere but right one is not

![2022-02-16_03 17 21](https://user-images.githubusercontent.com/53707587/154172460-d3442b8c-bc54-4b2c-a6cc-bd23086b6aff.png)

**Comment by maruohon** [comment]
Yep that is certainly weird when the center is at the corner of the blocks. I haven't properly debugged this yet, but it might be a sort of directional preference in the way the code builds the "rings" by walking around the perimeter. Maybe it's not properly trying all the required sides when it comes to the points where the "outside direction vector" kinda changes. I'll need to take a proper look at it at some point soon.

**Comment by HiggsBogan** [comment]
I was building a 128 radius sphere before i updated and now it doesnt line up no matter what i do. I keep thinking maybe i stuffed it up but theres no way i made this many mistakes when you can clearly see mis-placed blocks :(
![2022-03-05_06 14 38](https://user-images.githubusercontent.com/51397143/156827984-547f8cc7-478c-4695-87be-36e8146e662f.png)

**Comment by maruohon** [comment]
@HiggsBogan Is the center point, block snap setting and radius exactly the same as before? If you didn't use a block snap setting and you moved the center point even slightly, then it's possible that the positions changed like this. What version of the mod did you use before? If this was all in the 1.18+ versions, then *I don't think* I changed anything yet with the sphere stuff since the first "official" 1.18.0 release. I did rewrite some of the sphere stuff recently at one point, but I think it was leading up to the full 1.18.0 release.

I decided to leave the fixes this issue report is related to, to the bigger mod rewrites that are happening in 1.12.2, as I may need to or want to rewrite some more of how the shapes are generated, so I didn't want to delay the 1.18.2 updates further. So it's also possible that the sphere is just slightly crooked depending on the center point in relation to the block bounds (apparently). But even so it shouldn't change over time if the center and the settings remain the same.

**Comment by HiggsBogan** [comment]
not too long ago I messed up and deleted the sphere i was using. once i got it back to where it should be i took a screen shot of the coords and set up. the one above maybe off a bit because i was trying to get it right. just for you i re installed what i had before and set it to the settings in the screen shot i took for reference. heres that one
![2022-03-05_06 47 46](https://user-images.githubusercontent.com/51397143/156832256-a64b4113-6eca-47ef-a4b2-0bb87e7d0cda.png)

As you can see give or take the few mis placed blocks everything is fine.

Its all good :) I can wait or if i need to i will roll back when i need too. just thought id show you that its a bit wonky.

Thanks for your response

Oh also yeh I was using 0.19.0 - dev20211204.230657

**Comment by Firazoid** [comment]
Has there been any update on this bug? I am encountering the same issue. It seems to me like the misshapen side is always on the west, regardless of any combination of settings. I notice the images provided by the OP also show errors on the west side only.

As is, I consider the sphere tool to be unusable.

**Comment by Wolkenfarmer** [comment]
Still present in 1.19.2.

**Comment by Jescudie** [comment]
Still present in 1.20.1.

**Comment by obfuscoder** [comment]
Before creating my issue, I failed to use the search function to figure out if this was already reported. I had the same issue wiuth spheres. It does not matter what snap method is used. The sphere is always asymetrical. Cylinders shapes ARE symetrical so my guess is that the function to calculate the points on the sphere is faulty (My guess is that the floating point arythmetic is incorrect. Ssome rounding issue? Using integers where floating/fixed point values ought to be used?)..

**Comment by OldMensch** [comment]
Still an issue in 1.20.4

**Comment by Alexandre-LvLDesign** [comment]
![seriouslybro](https://github.com/maruohon/minihud/assets/39302412/7b99d54d-79d0-4022-8dd9-e46f7f8de1da)
Still an issue in 1.20.4 ( 29/02/2024) 

the bug is present since Feb 2022, I don't think it will be fix

**Comment by CaduceusBasilisk** [comment]
Still an issue in 1.20.4 (03/05/2024)
![IMG_5848](https://github.com/maruohon/minihud/assets/168820751/a1b3fbd6-e81a-4f3f-9a72-a498aa23414a)

**Comment by valve-developer** [comment]
![Untitled](https://github.com/maruohon/minihud/assets/158013759/1826ab9b-875a-4709-9a20-1879e1981a10)
still present in 1.21

**Comment by sakura-ryoko** [comment]
I am aware of it, I just haven't had a chance to look at it yet (1.20.6/1.21)

**Comment by AdielHub** [comment]
I guess in the meantime the only fix is to make a cylinder with a radius of +1/-1 for each increment of y level

Very very very inconvenient but it's better than a wrong sphere


Maruohon, like, disable the sphere feature until it's fixed. 
The only thing worse than not having a sphere tool is having a sphere tool that you only realise hasn't worked properly until your minecraft project is done

**Comment by Hlebuw3k** [comment]
I built something with this, unaware that it was not correct. Please disable the feature until fixed.

**Comment by Caderk** [comment]
> I guess in the meantime the only fix is to make a cylinder with a radius of +1/-1 for each increment of y level

That's a cone my brother, not a sphere.

**Comment by JamesSG2** [comment]
This issue still persists in Java 1.21.4! (Latest Minihud as of 12/13/24) Clearly these are not perfect spheres even when using the sphere shape in Minihud. They also don't agree with: https://www.plotz.co.uk/plotz-model.php?model=Sphere which I know has correct dimensions. Please fix this stat! Seems like the most basic feature of Minihud doesn't work correctly:

![2024-12-14_21 33 29](https://github.com/user-attachments/assets/3fa04cfd-2ba4-431d-99eb-6bb365f2480c)
![2024-12-14_21 43 01](https://github.com/user-attachments/assets/14422c86-31aa-4f04-9b19-f37c96d80dbe)

**Comment by adam2-a** [comment]
This issue is still present in Java 1.21.4 (using the latest version of MiniHUD as of 12/13/24). The spheres generated with the "sphere" shape in MiniHUD are clearly not true spheres—they don’t even match the dimensions shown on [Plotz Sphere Generator](https://minecraftcirclesgenerator.com/), which I know is accurate. Please address this ASAP—this seems like one of the most basic features, and it’s not working as expected.

**Comment by YourMJK** [comment]
@JamesSG2 @adam2-a  
After finding this issue, I wanted to fix it with a PR but (un)fortunately I can't replicate the wonky spheres in 1.21.10 with the latest version of MiniHUD (0.37.2).

Even when I use the exact same coordinates and settings as in @scrapmech's original screenshot, I still get a perfectly symmetrical sphere from all sides.

Was this fixed?  
Or can someone give me a reproducible configuration?

**Comment by sakura-ryoko** [comment]
> [@JamesSG2](https://github.com/JamesSG2) [@adam2-a](https://github.com/adam2-a) After finding this issue, I wanted to fix it with a PR but (un)fortunately I can't replicate the wonky spheres in 1.21.10 with the latest version of MiniHUD (0.37.2).
> 
> Even when I use the exact same coordinates and settings as in [@scrapmech](https://github.com/scrapmech)'s original screenshot, I still get a perfectly symmetrical sphere from all sides.
> 
> Was this fixed? Or can someone give me a reproducible configuration?

There was already a PR .

**Comment by sakura-ryoko** [comment]
SEE: https://github.com/sakura-ryoko/minihud/commit/5ed184a2d6abe432a59e17e1d0fc23bee177fb8d
